velox/functions/prestosql/HyperLogLogFunctions.h

```diff
@@ -78,4 +78,149 @@ struct EmptyApproxSetWithMaxErrorFunction {
+    auto arrayElements = baseArray->elements();
+    auto flatArrayElements = arrayElements->as<FlatVector<StringView>>();
+
+    memory::MemoryPool* memoryPool = context.pool();
+    using SparseHll = common::hll::SparseHll<memory::MemoryPool>;
+    using DenseHll = common::hll::DenseHll<memory::MemoryPool>;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (decodedArray->isNullAt(row)) {
+        flatResult->setNull(row, true);
+        return;
+      }
+
+      auto arraySize = rawSizes[indices[row]];
+      auto arrayOffset = rawOffsets[indices[row]];
+
+      int8_t indexBitLength = -1;
+      bool isSparse = true;
+      std::optional<SparseHll> sparseHll;
+      std::optional<DenseHll> denseHll;
+
+      for (vector_size_t i = 0; i < arraySize; ++i) {
+        auto elementIndex = arrayOffset + i;
+        if (arrayElements->isNullAt(elementIndex)) {
+          continue;
+        }
+
+        auto hllData = flatArrayElements->valueAt(elementIndex);
+        if (hllData.empty()) {
+          continue;
+        }
+
+        bool isSparseHllData = false;
+        bool isDenseHllData = false;
+
+        if (indexBitLength < 0) {
+          if (common::hll::DenseHlls::canDeserialize(hllData.data())) {
+            isDenseHllData = true;
+            indexBitLength = common::hll::DenseHlls::deserializeIndexBitLength(
+                hllData.data());
+          } else if (common::hll::SparseHlls::canDeserialize(hllData.data())) {
+            isSparseHllData = true;
+            indexBitLength = common::hll::SparseHlls::deserializeIndexBitLength(
+                hllData.data());
+          }
+        } else {
+          isSparseHllData =
+              common::hll::SparseHlls::canDeserialize(hllData.data());
+          isDenseHllData = !isSparseHllData &&
+              common::hll::DenseHlls::canDeserialize(hllData.data());
+        }
+
+        if (isSparseHllData) {
+          auto cardinality =
+              common::hll::SparseHlls::cardinality(hllData.data());
+          if (cardinality == 0) {
+            continue;
+          }
+
+          if (isSparse) {
+            if (!sparseHll) {
+              sparseHll.emplace(memoryPool);
+              sparseHll->setSoftMemoryLimit(
+                  common::hll::DenseHlls::estimateInMemorySize(indexBitLength));
+            }
+
+            sparseHll->mergeWith(hllData.data());
+            if (sparseHll->overLimit()) {
+              denseHll.emplace(indexBitLength, memoryPool);
+              sparseHll->toDense(*denseHll);
+              sparseHll.reset();
+              isSparse = false;
+            }
+          } else {
+            SparseHll temp{hllData.data(), memoryPool};
+            temp.toDense(*denseHll);
+          }
+        } else if (common::hll::DenseHlls::canDeserialize(hllData.data())) {
+          auto hllIndexBitLength =
+              common::hll::DenseHlls::deserializeIndexBitLength(hllData.data());
+          VELOX_USER_CHECK_EQ(
```

Comment:
Vector functions should not throw. See https://facebookincubator.github.io/velox/develop/scalar-functions.html#try-expression-support

Please, fix and add test coverage.

Reply:
Thanks for the link. Very useful. Updated to applyToSelectedNoThrow() instead of applyToSelected() does that make sense?